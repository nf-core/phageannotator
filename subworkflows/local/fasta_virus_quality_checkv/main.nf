//
// Assess virus quality with Checkv
//

include { CHECKV_DOWNLOADDATABASE   } from '../../../modules/nf-core/checkv/downloaddatabase/main'
include { UNTAR                     } from '../../../modules/nf-core/untar/main'
include { CHECKV_ENDTOEND           } from '../../../modules/nf-core/checkv/endtoend/main'          // TODO: Update nf-core module to gzip output FASTA files

workflow FASTA_VIRUS_QUALITY_CHECKV {
    take:
    virus_fasta_gz  // [ [ meta ], fasta.gz ]   , assemblies/genomes (mandatory)
    checkv_db       // [ checkv_db ]            , CheckV database directory (optional)

    main:
    ch_versions = Channel.empty()

    // if genomad_db exists, skip CHECKV_DOWNLOADDATABASE
    if ( checkv_db ){
        ch_checkv_db = checkv_db
    } else {
        // download standard checkv db unless minimal db is specified
        if ( !params.checkv_minimal_db ){
            //
            // MODULE: download standard CheckV database
            //
            ch_checkv_db = CHECKV_DOWNLOADDATABASE( ).checkv_db
            ch_versions = ch_versions.mix(CHECKV_DOWNLOADDATABASE.out.versions)
        } else {
            //
            // MODULE: untar minimal checkv database
            //
            ch_checkv_db = UNTAR ( [ [ id:'checkv_minimal_db' ], file("https://github.com/nf-core/test-datasets/raw/phageannotator/modules/nfcore/checkv/endtoend/checkv_minimal_db.tar", checkIfExists: true) ] ).untar.map { it[1] }
            ch_versions = ch_versions.mix(UNTAR.out.versions)
        }
    }

    //
    // MODULE: Assess virus quality
    //
    CHECKV_ENDTOEND ( virus_fasta_gz, ch_checkv_db.collect() )
    ch_quality_summary_tsv  = CHECKV_ENDTOEND.out.quality_summary
    ch_viruses_fna_gz       = CHECKV_ENDTOEND.out.viruses
    ch_proviruses_fna_gz    = CHECKV_ENDTOEND.out.proviruses
    ch_versions = ch_versions.mix(CHECKV_ENDTOEND.out.versions.first())

    emit:
    viruses_fna_gz      = ch_viruses_fna_gz         // [ [ meta ], viruses.fna.gz ]       , FASTA file containing viruses
    proviruses_fna_gz   = ch_proviruses_fna_gz      // [ [ meta ], proviruses.fna.gz ]    , FASTA file containing proviruses
    quality_summary_tsv = ch_quality_summary_tsv    // [ [ meta ], quality_summary.tsv ]  , TSV file containing quality data
    versions            = ch_versions               // [ versions.yml ]
}
