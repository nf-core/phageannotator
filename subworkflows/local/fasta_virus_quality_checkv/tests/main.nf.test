nextflow_workflow {

    name "Test Subworkflow: FASTA_VIRUS_QUALITY_CHECKV"
    script "subworkflows/local/fasta_virus_quality_checkv/main.nf"
    workflow "FASTA_VIRUS_QUALITY_CHECKV"

    // Dependencies
    tag "CHECKV_DOWNLOADDATABASE"
    tag "UNTAR"
    tag "CHECKV_ENDTOEND"

    test("fasta.gz") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ],
                        file(params.modules_testdata_base_path + 'genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz', checkIfExists: true)
                    ]
                )
                input[1] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
