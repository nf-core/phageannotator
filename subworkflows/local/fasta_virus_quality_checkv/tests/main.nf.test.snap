{
    "fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2"
                        },
                        "viruses.fna.gz:md5,67028c9f9447855f86013fd43a5f6105"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "viruses.fna.gz:md5,67028c9f9447855f86013fd43a5f6105"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test2"
                        },
                        "proviruses.fna.gz:md5,f2b8287ef49757e3a0320217b8add37d"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "proviruses.fna.gz:md5,f2b8287ef49757e3a0320217b8add37d"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test2"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ]
                ],
                "3": [
                    "versions.yml:md5,58c018640348d4755b08dd850a477465",
                    "versions.yml:md5,bde8142f045061369c16cd56d51698ed"
                ],
                "proviruses_fna_gz": [
                    [
                        {
                            "id": "test2"
                        },
                        "proviruses.fna.gz:md5,f2b8287ef49757e3a0320217b8add37d"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "proviruses.fna.gz:md5,f2b8287ef49757e3a0320217b8add37d"
                    ]
                ],
                "quality_summary_tsv": [
                    [
                        {
                            "id": "test2"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "quality_summary.tsv:md5,04e827b7efefe52727a904cfd3af1639"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,58c018640348d4755b08dd850a477465",
                    "versions.yml:md5,bde8142f045061369c16cd56d51698ed"
                ],
                "viruses_fna_gz": [
                    [
                        {
                            "id": "test2"
                        },
                        "viruses.fna.gz:md5,67028c9f9447855f86013fd43a5f6105"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "viruses.fna.gz:md5,67028c9f9447855f86013fd43a5f6105"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-28T16:47:23.174265397"
    }
}
